### PR TITLE
 12-factor app using devon4j

### PR DESCRIPTION
**Pull request by ssarmokadam**
guide on how to create 12-factor app with devon4j

### CODE REVIEW

File: documentation/12-factor-app-devon4j.asciidoc [hunk 1]
```diff
@@ -0,0 +1,151 @@
...
+This document mainly focuses on discussing how can you create 12 factor app with devon4j. To know more about this 12 factors you can refer https://12factor.net/[here] . Twelve factor is mainly focus on creating cloud native applications. These are the guidelines on what factors you need to consider in different stages of application lifecycle.
+
+== Codebase
+
+*One codebase tracked in revision control, many deploys*
+
+First factor to consider while creating cloud native application is 1 codebase = 1 application = 1 process. 
+
+Traditionally we are used to develop and maintain monolithic applications. But it is difficult to deploy monolith appication as it is into the cloud environment it might night give us a benefit of cloud if we just do lift and shift. So we need to change our application architecture to adapt it like a cloud applications. One of the way to consider is breaking your application into multiple microservices. For simple application it might not be applicable but for some complex system user can check if they want to go ahead with microservices approach.
+To satisfy this factor consider creating one git/code repository for 1 process.
```
Comment (SchettlerKoehler) [inline documentation/12-factor-app-devon4j.asciidoc:17]: for each process

File: documentation/12-factor-app-devon4j.asciidoc [hunk 2]
```diff
@@ -0,0 +1,151 @@
...
+
+== Dependencies
+
+*Explicitly declare and isolate dependencies*
+
+A twelve-factor app never relies on implicit existence of system-wide packages. It should explicitly declare and isolate dependencies. This is required to make repeatable deployments of an app without building it again and again.
+
+devon4j application is a maven project so we declare dependencies in pom.xml. For isolation of dependencies we should containerize an application.It will isolate application environment from rest of the environment.
```
Comment (SchettlerKoehler) [inline documentation/12-factor-app-devon4j.asciidoc:25]: Here you are missing a space between the last sentences.

Also, we should add a link on how to put the application in a container. For both quarkus and spring

File: documentation/12-factor-app-devon4j.asciidoc [hunk 3]
```diff
@@ -0,0 +1,151 @@
...
+
+== Config
+
+*Store config in the environment*
+
+What is configuration?
+
+* Resource handles to database or other backing services
+* Credential to external sources
+* Anything that will vary between different deploys like in dev, test, staging and prod.
```
Comment (SchettlerKoehler) [inline documentation/12-factor-app-devon4j.asciidoc:35]: deployments or better deployment environments

File: documentation/12-factor-app-devon4j.asciidoc [hunk 4]
```diff
@@ -0,0 +1,151 @@
...
+
+Never ever hardcode this configurations into code or do not put it in config files. This is because config files are also bundled with source code and this values will change depending on which environment you want to deploy your application.
```
Comment (SchettlerKoehler) [inline documentation/12-factor-app-devon4j.asciidoc:37]: or -> and

File: documentation/12-factor-app-devon4j.asciidoc [hunk 5]
```diff
@@ -0,0 +1,151 @@
...
+*Store config in the environment*
+
+What is configuration?
+
+* Resource handles to database or other backing services
+* Credential to external sources
+* Anything that will vary between different deploys like in dev, test, staging and prod.
+
+Never ever hardcode this configurations into code or do not put it in config files. This is because config files are also bundled with source code and this values will change depending on which environment you want to deploy your application. 
+This factor has to be considered while designing/implementing your project. Consider reading actual value of property from environment variables.
```
Comment (SchettlerKoehler) [inline documentation/12-factor-app-devon4j.asciidoc:38]: We should have a guide for that

File: documentation/12-factor-app-devon4j.asciidoc [hunk 6]
```diff
@@ -0,0 +1,151 @@
...
+
+[source]
+----
+logging:
+  level:
+    org.springframework: ${SPRING_LOG_LEVEL:INFO}
+    hello: ${LOG_LEVEL:INFO}
+----
+
+As shown in above image, default value of logging level is INFO but in case I want to change it to debug in test environment to analyze some issue then I do not need to change it into the code or in yaml file. Instead I can change value of environment variable SPRING_LOG_LEVEL to debug.
```
Comment (SchettlerKoehler) [inline documentation/12-factor-app-devon4j.asciidoc:49]: image -> code block or code

Comment (SchettlerKoehler) [reply documentation/12-factor-app-devon4j.asciidoc:49]: You should point out that every variable should have a reasonable default

**Review by SchettlerKoehler — CHANGES_REQUESTED** [review]
First comments. Rest will follow

**Review by maybeec — COMMENTED** [review]
Rather than putting a lot of text here and trying to re-explain 12-factors, I would strongly encourage you to introduce 12-factors with only 1-2 sentences referring to 12factor.net for further reading and reading the explanation of the 12 factors.

In addition, you could have a table or even chapters for each factor simply with one sentence referring to some documentation or solution or others. We should not try to re-explain what is understood from the factor, but rather point to the guidelines how devonfw is handling each factor. This could even be multiple different implementations depending on the programming language.

File: documentation/12-factor-app-devon4j.asciidoc [hunk 7]
```diff
@@ -0,0 +1,96 @@
+:toc: macro
+toc::[]
+:idprefix:
+:idseparator: -
+
+= 12-factor-app with devon4j
```
Comment (hohwille) [inline documentation/12-factor-app-devon4j.asciidoc:6]: The images are nowhere referenced.
Further `12-factor-app-config.JPG` only contains source-code that can be included directly as text into asciidoc.

**Review by SchettlerKoehler — APPROVED** [review]

**Pull request merged**